File: src/nix-protocol/hs/test/Captures.hs
-- SPDX-License-Identifier: MIT
-- | Test Haskell serializers against captured binary data

module Main (main) where

import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import Data.Word (Word64)
import Nix.Protocol
import System.FilePath ((</>))
import Test.Tasty
import Test.Tasty.HUnit

-- | Directory containing captures (relative to test execution)
capturesDir :: FilePath
capturesDir = "captures"

readCapture :: FilePath -> IO ByteString
readCapture name = BS.readFile (capturesDir </> name)

-- | Parse a string from captured binary data at given offset
parseString :: ByteString -> Int -> (Text, Int)
parseString bs offset =
  let len = fromIntegral $ readWord64LE bs offset
      str = TE.decodeUtf8 $ BS.take len $ BS.drop (offset + 8) bs
      padding = (8 - (len `mod` 8)) `mod` 8
   in (str, offset + 8 + len + padding)

-- | Read a little-endian Word64 from ByteString
readWord64LE :: ByteString -> Int -> Word64
readWord64LE bs offset =
  let bytes = BS.unpack $ BS.take 8 $ BS.drop offset bs
   in foldr (\(i, b) acc -> acc + fromIntegral b * (256 ^ i)) 0 (zip [0 ..] bytes)

-- =============================================================================
-- Tests
-- =============================================================================

main :: IO ()
main = defaultMain tests

tests :: TestTree
tests =
  testGroup
    "Captures"
    [ testGroup
        "Writer"
        [ testClientHello,
          testServerHello,
          testIsValidPathRequest,
          testQueryPathInfoRequest,
          testQueryReferrersRequest,
          testAddTempRootRequest,
          testAddIndirectRootRequest,
          testFindRootsRequest,
          testNarFromPathRequest,
          testQueryMissingRequest,
          testBuildPathsRequest,
          testBuildPathsWithResultsRequest,
          testSetOptionsRequest,
          testAddToStoreNarRequest
        ],
      testGroup
        "Reader"
        [ testReadServerHello,
          testReadIsValidPathResponse,
          testReadQueryPathInfoResponse,
          testReadQueryMissingResponse,
          testReadQueryReferrersResponse,
          testReadBuildPathsWithResultsResponse
        ]
    ]

testClientHello :: TestTree
testClientHello = testCase "client_hello" $ do
  let generated = execWriter $ writeClientHello 0x0126
  expected <- readCapture "client_hello.bin"
  generated @?= expected

testServerHello :: TestTree
testServerHello = testCase "server_hello" $ do
  let generated = execWriter $ writeServerHello 0x0126
  expected <- readCapture "server_hello.bin"
  generated @?= expected

testIsValidPathRequest :: TestTree
testIsValidPathRequest = testCase "isvalidpath_request" $ do
  captured <- readCapture "isvalidpath_request.bin"
  let (path, _) = parseString captured 8
      generated = execWriter $ writeIsValidPathRequest path
  generated @?= captured

testQueryPathInfoRequest :: TestTree
testQueryPathInfoRequest = testCase "querypathinfo_request" $ do
  captured <- readCapture "querypathinfo_request.bin"
  let (path, _) = parseString captured 8
      generated = execWriter $ writeQueryPathInfoRequest path
  generated @?= captured

testQueryReferrersRequest :: TestTree
testQueryReferrersRequest = testCase "queryreferrers_request" $ do
  captured <- readCapture "queryreferrers_request.bin"
  let (path, _) = parseString captured 8
      generated = execWriter $ writeQueryReferrersRequest path
  generated @?= captured

testAddTempRootRequest :: TestTree
testAddTempRootRequest = testCase "addtemproot_request" $ do
  captured <- readCapture "addtemproot_request.bin"
  let (path, _) = parseString captured 8
      generated = execWriter $ writeAddTempRootRequest path
  generated @?= captured

testAddIndirectRootRequest :: TestTree
testAddIndirectRootRequest = testCase "addindirectroot_request" $ do
  captured <- readCapture "addindirectroot_request.bin"
  let (path, _) = parseString captured 8
      generated = execWriter $ writeAddIndirectRootRequest path
  generated @?= captured

testFindRootsRequest :: TestTree
testFindRootsRequest = testCase "findroots_request" $ do
  let generated = execWriter writeFindRootsRequest
  expected <- readCapture "findroots_request.bin"
  generated @?= expected

testNarFromPathRequest :: TestTree
testNarFromPathRequest = testCase "narfrompath_request" $ do
  captured <- readCapture "narfrompath_request.bin"
  let (path, _) = parseString captured 8
      generated = execWriter $ writeNarFromPathRequest path
  generated @?= captured

testQueryMissingRequest :: TestTree
testQueryMissingRequest = testCase "querymissing_request" $ do
  captured <- readCapture "querymissing_request.bin"
  let numPaths = fromIntegral $ readWord64LE captured 8 :: Int
      parsePaths :: Int -> Int -> [Text] -> ([Text], Int)
      parsePaths 0 offset acc = (acc, offset)
      parsePaths n offset acc =
        let (path, newOffset) = parseString captured offset
         in parsePaths (n - 1) newOffset (acc ++ [path])
      (paths, _) = parsePaths numPaths 16 []
      generated = execWriter $ writeQueryMissingRequest paths
  generated @?= captured

testBuildPathsRequest :: TestTree
testBuildPathsRequest = testCase "buildpaths_request" $ do
  captured <- readCapture "buildpaths_request.bin"
  let numPaths = fromIntegral $ readWord64LE captured 8
      parsePaths 0 offset acc = (acc, offset)
      parsePaths n offset acc =
        let (path, newOffset) = parseString captured offset
         in parsePaths (n - 1) newOffset (acc ++ [path])
      (paths, modeOffset) = parsePaths numPaths 16 []
      modeVal = readWord64LE captured modeOffset
      mode = case modeVal of
        0 -> Normal
        1 -> Repair
        2 -> Check
        _ -> error "Unknown build mode"
      generated = execWriter $ writeBuildPathsRequest paths mode
  generated @?= captured

testBuildPathsWithResultsRequest :: TestTree
testBuildPathsWithResultsRequest = testCase "buildpathswithresults_request" $ do
  captured <- readCapture "buildpathswithresults_request.bin"
  let numPaths = fromIntegral $ readWord64LE captured 8
      parsePaths 0 offset acc = (acc, offset)
      parsePaths n offset acc =
        let (path, newOffset) = parseString captured offset
         in parsePaths (n - 1) newOffset (acc ++ [path])
      (paths, modeOffset) = parsePaths numPaths 16 []
      modeVal = readWord64LE captured modeOffset
      mode = case modeVal of
        0 -> Normal
        1 -> Repair
        2 -> Check
        _ -> error "Unknown build mode"
      generated = execWriter $ writeBuildPathsWithResultsRequest paths mode
  generated @?= captured

testSetOptionsRequest :: TestTree
testSetOptionsRequest = testCase "setoptions_request" $ do
  let settings =
        ClientSettings
          { csKeepFailed = False,
            csKeepGoing = False,
            csTryFallback = False,
            csVerbosity = 3,
            csMaxBuildJobs = 32,
            csMaxSilentTime = 0,
            csUseBuildHook = True,
            csVerboseBuild = 7,
            csLogType = 0,
            csPrintBuildTrace = 0,
            csBuildCores = 0,
            csUseSubstitutes = True,
            csOverrides =
              [ (T.pack "extra-platforms", T.pack "aarch64-linux"),
                (T.pack "sandbox", T.pack "false"),
                (T.pack "substituters", T.pack "https://cache.nixos.org https://weyl-ai.cachix.org"),
                ( T.pack "trusted-public-keys",
                  T.pack $
                    "cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY= "
                      ++ "weyl-ai.cachix.org-1:cR0SpSAPw7wejZ21ep4SLojE77gp5F2os260eEWqTTw="
                )
              ]
          }
      generated = execWriter $ writeSetOptionsRequest settings 38
  expected <- readCapture "setoptions_request.bin"
  generated @?= expected

testAddToStoreNarRequest :: TestTree
testAddToStoreNarRequest = testCase "addtostorenar_request" $ do
  let req =
        AddToStoreNarRequest
          { asnPath = T.pack "/nix/store/v4wqkf7dq9619yyv1wbf1jvpw4dhbs2f-tf.txt",
            asnDeriver = T.empty,
            asnNarHash = T.pack "60e5106c2a1d4ef9f82b58df02be7a482f35438ff9e6556194ab02e355256352",
            asnReferences = [],
            asnRegistrationTime = 0,
            asnNarSize = 136,
            asnUltimate = False,
            asnSignatures = [],
            asnCa = T.pack "fixed:sha256:081b9nklhcc0gck2a6j821xij57djvnfxg1fps0f5l2bbpd1sgp1",
            asnRepair = False,
            asnDontCheckSigs = False
          }
      generated = execWriter $ writeAddToStoreNarRequest req
  expected <- readCapture "addtostorenar_request.bin"
  generated @?= expected

-- =============================================================================
-- Reader Tests
-- =============================================================================

testReadServerHello :: TestTree
testReadServerHello = testCase "read_server_hello" $ do
  captured <- readCapture "server_hello.bin"
  case runReader readServerHello captured of
    Left err -> assertFailure $ "Parse failed: " ++ err
    Right hello -> do
      shMagic hello @?= workerMagic2
      shVersion hello @?= 0x0126

testReadIsValidPathResponse :: TestTree
testReadIsValidPathResponse = testCase "read_isvalidpath_response" $ do
  captured <- readCapture "isvalidpath_response.bin"
  case runReader readIsValidPathResponse captured of
    Left err -> assertFailure $ "Parse failed: " ++ err
    Right valid -> valid @?= True

testReadQueryPathInfoResponse :: TestTree
testReadQueryPathInfoResponse = testCase "read_querypathinfo_response" $ do
  captured <- readCapture "querypathinfo_response.bin"
  case runReader (readQueryPathInfoResponse 0x0126) captured of
    Left err -> assertFailure $ "Parse failed: " ++ err
    Right Nothing -> assertFailure "Expected Some, got Nothing"
    Right (Just info) -> do
      assertBool "deriver contains bash" $ T.isInfixOf (T.pack "bash") (vpiDeriver info)
      assertBool "nar_hash starts with f7b02ee0" $ T.isPrefixOf (T.pack "f7b02ee0") (vpiNarHash info)
      length (vpiReferences info) @?= 2
      length (vpiSignatures info) @?= 1

testReadQueryMissingResponse :: TestTree
testReadQueryMissingResponse = testCase "read_querymissing_response" $ do
  captured <- readCapture "querymissing_response.bin"
  case runReader readQueryMissingResponse captured of
    Left err -> assertFailure $ "Parse failed: " ++ err
    Right result -> do
      null (qmrWillBuild result) @?= True
      null (qmrWillSubstitute result) @?= True
      qmrDownloadSize result @?= 0
      qmrNarSize result @?= 0

testReadQueryReferrersResponse :: TestTree
testReadQueryReferrersResponse = testCase "read_queryreferrers_response" $ do
  captured <- readCapture "queryreferrers_response.bin"
  case runReader readQueryReferrersResponse captured of
    Left err -> assertFailure $ "Parse failed: " ++ err
    Right refs -> do
      assertBool "not empty" $ not (null refs)
      assertBool "first is store path" $ T.isPrefixOf (T.pack "/nix/store/") (head refs)

testReadBuildPathsWithResultsResponse :: TestTree
testReadBuildPathsWithResultsResponse = testCase "read_buildpathswithresults_response" $ do
  captured <- readCapture "buildpathswithresults_response.bin"
  case runReader (readBuildPathsWithResultsResponse 0x0126) captured of
    Left err -> assertFailure $ "Parse failed: " ++ err
    Right results -> do
      length results @?= 1
      let result = head results
      assertBool "path contains hello" $ T.isInfixOf (T.pack "hello") (brpPath result)
      brpStatus result @?= 2
      length (brpBuiltOutputs result) @?= 1

File: src/nix-protocol/tests/captures.rs
// SPDX-License-Identifier: MIT
//! Test Rust serializers against captured binary data

use nix_protocol::*;
use std::fs;
use std::path::PathBuf;

fn captures_dir() -> PathBuf {
  PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("captures")
}

fn read_capture(name: &str) -> Vec<u8> {
  let path = captures_dir().join(name);
  fs::read(&path).unwrap_or_else(|e| panic!("Failed to read {}: {}", path.display(), e))
}

fn compare_buffers(generated: &[u8], expected: &[u8], name: &str) {
  assert_eq!(
    generated.len(),
    expected.len(),
    "{}: size mismatch (got {}, expected {})",
    name,
    generated.len(),
    expected.len()
  );

  for (i, (g, e)) in generated.iter().zip(expected.iter()).enumerate() {
    assert_eq!(
      g, e,
      "{}: byte mismatch at offset {} (got 0x{:02x}, expected 0x{:02x})",
      name, i, g, e
    );
  }
}

// Helper to parse a string from captured binary data
fn parse_string(data: &[u8], offset: usize) -> (String, usize) {
  let len = u64::from_le_bytes(data[offset..offset + 8].try_into().unwrap()) as usize;
  let s = String::from_utf8(data[offset + 8..offset + 8 + len].to_vec()).unwrap();
  let padding = (8 - (len % 8)) % 8;
  (s, offset + 8 + len + padding)
}

// =============================================================================
// Tests
// =============================================================================

#[test]
fn test_client_hello() {
  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_client_hello(&mut w, 0x0126).unwrap();

  let expected = read_capture("client_hello.bin");
  compare_buffers(&buf, &expected, "client_hello");
}

#[test]
fn test_server_hello() {
  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_server_hello(&mut w, 0x0126).unwrap();

  let expected = read_capture("server_hello.bin");
  compare_buffers(&buf, &expected, "server_hello");
}

#[test]
fn test_isvalidpath_request() {
  let captured = read_capture("isvalidpath_request.bin");
  let (path, _) = parse_string(&captured, 8);

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_is_valid_path_request(&mut w, &path).unwrap();

  compare_buffers(&buf, &captured, "isvalidpath_request");
}

#[test]
fn test_querypathinfo_request() {
  let captured = read_capture("querypathinfo_request.bin");
  let (path, _) = parse_string(&captured, 8);

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_query_path_info_request(&mut w, &path).unwrap();

  compare_buffers(&buf, &captured, "querypathinfo_request");
}

#[test]
fn test_queryreferrers_request() {
  let captured = read_capture("queryreferrers_request.bin");
  let (path, _) = parse_string(&captured, 8);

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_query_referrers_request(&mut w, &path).unwrap();

  compare_buffers(&buf, &captured, "queryreferrers_request");
}

#[test]
fn test_addtemproot_request() {
  let captured = read_capture("addtemproot_request.bin");
  let (path, _) = parse_string(&captured, 8);

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_add_temp_root_request(&mut w, &path).unwrap();

  compare_buffers(&buf, &captured, "addtemproot_request");
}

#[test]
fn test_addindirectroot_request() {
  let captured = read_capture("addindirectroot_request.bin");
  let (path, _) = parse_string(&captured, 8);

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_add_indirect_root_request(&mut w, &path).unwrap();

  compare_buffers(&buf, &captured, "addindirectroot_request");
}

#[test]
fn test_findroots_request() {
  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_find_roots_request(&mut w).unwrap();

  let expected = read_capture("findroots_request.bin");
  compare_buffers(&buf, &expected, "findroots_request");
}

#[test]
fn test_narfrompath_request() {
  let captured = read_capture("narfrompath_request.bin");
  let (path, _) = parse_string(&captured, 8);

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_nar_from_path_request(&mut w, &path).unwrap();

  compare_buffers(&buf, &captured, "narfrompath_request");
}

#[test]
fn test_querymissing_request() {
  let captured = read_capture("querymissing_request.bin");

  // Parse: op(8) + num_paths(8) + paths...
  let num_paths = u64::from_le_bytes(captured[8..16].try_into().unwrap()) as usize;
  let mut paths = Vec::new();
  let mut offset = 16;

  for _ in 0..num_paths {
    let (path, new_offset) = parse_string(&captured, offset);
    paths.push(path);
    offset = new_offset;
  }

  let path_refs: Vec<&str> = paths.iter().map(|s| s.as_str()).collect();

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_query_missing_request(&mut w, &path_refs).unwrap();

  compare_buffers(&buf, &captured, "querymissing_request");
}

#[test]
fn test_buildpaths_request() {
  let captured = read_capture("buildpaths_request.bin");

  // Parse: op(8) + num_paths(8) + paths... + mode(8)
  let num_paths = u64::from_le_bytes(captured[8..16].try_into().unwrap()) as usize;
  let mut paths = Vec::new();
  let mut offset = 16;

  for _ in 0..num_paths {
    let (path, new_offset) = parse_string(&captured, offset);
    paths.push(path);
    offset = new_offset;
  }

  let mode_val = u64::from_le_bytes(captured[offset..offset + 8].try_into().unwrap());
  let mode = match mode_val {
    0 => BuildMode::Normal,
    1 => BuildMode::Repair,
    2 => BuildMode::Check,
    _ => panic!("Unknown build mode: {}", mode_val),
  };

  let path_refs: Vec<&str> = paths.iter().map(|s| s.as_str()).collect();

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_build_paths_request(&mut w, &path_refs, mode).unwrap();

  compare_buffers(&buf, &captured, "buildpaths_request");
}

#[test]
fn test_buildpathswithresults_request() {
  let captured = read_capture("buildpathswithresults_request.bin");

  // Parse: op(8) + num_paths(8) + paths... + mode(8)
  let num_paths = u64::from_le_bytes(captured[8..16].try_into().unwrap()) as usize;
  let mut paths = Vec::new();
  let mut offset = 16;

  for _ in 0..num_paths {
    let (path, new_offset) = parse_string(&captured, offset);
    paths.push(path);
    offset = new_offset;
  }

  let mode_val = u64::from_le_bytes(captured[offset..offset + 8].try_into().unwrap());
  let mode = match mode_val {
    0 => BuildMode::Normal,
    1 => BuildMode::Repair,
    2 => BuildMode::Check,
    _ => panic!("Unknown build mode: {}", mode_val),
  };

  let path_refs: Vec<&str> = paths.iter().map(|s| s.as_str()).collect();

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_build_paths_with_results_request(&mut w, &path_refs, mode).unwrap();

  compare_buffers(&buf, &captured, "buildpathswithresults_request");
}

#[test]
fn test_setoptions_request() {
  // Hardcoded values from captured traffic
  let settings = ClientSettings {
    keep_failed: false,
    keep_going: false,
    try_fallback: false,
    verbosity: 3,
    max_build_jobs: 32,
    max_silent_time: 0,
    use_build_hook: true,
    verbose_build: 7,
    log_type: 0,
    print_build_trace: 0,
    build_cores: 0,
    use_substitutes: true,
    overrides: vec![
      ("extra-platforms".to_string(), "aarch64-linux".to_string()),
      ("sandbox".to_string(), "false".to_string()),
      (
        "substituters".to_string(),
        "https://cache.nixos.org https://weyl-ai.cachix.org".to_string(),
      ),
      (
        "trusted-public-keys".to_string(),
        "cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY= \
         weyl-ai.cachix.org-1:cR0SpSAPw7wejZ21ep4SLojE77gp5F2os260eEWqTTw="
          .to_string(),
      ),
    ],
  };

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_set_options_request(&mut w, &settings, 38).unwrap();

  let expected = read_capture("setoptions_request.bin");
  compare_buffers(&buf, &expected, "setoptions_request");
}

#[test]
fn test_addtostorenar_request() {
  let req = AddToStoreNarRequest {
    path: "/nix/store/v4wqkf7dq9619yyv1wbf1jvpw4dhbs2f-tf.txt".to_string(),
    deriver: String::new(),
    nar_hash: "60e5106c2a1d4ef9f82b58df02be7a482f35438ff9e6556194ab02e355256352".to_string(),
    references: vec![],
    registration_time: 0,
    nar_size: 136,
    ultimate: false,
    signatures: vec![],
    ca: "fixed:sha256:081b9nklhcc0gck2a6j821xij57djvnfxg1fps0f5l2bbpd1sgp1".to_string(),
    repair: false,
    dont_check_sigs: false,
  };

  let mut buf = Vec::new();
  let mut w = Writer::new(&mut buf);
  write_add_to_store_nar_request(&mut w, &req).unwrap();

  let expected = read_capture("addtostorenar_request.bin");
  compare_buffers(&buf, &expected, "addtostorenar_request");
}
